### ISSUE TITLE
Configure "solaris64-sparcv9-gcc" but shows "THIRTY_TWO_BIT mode" in 1.1.0c

### ISSUE THREAD

**Issue by dyduyu** [openssl/openssl#2197]
Hi,

  I'm trying to build 64 bit library(openssl 1.1.0c) on solaris sparc(5.11), I specify **solaris64-sparcv9-gcc** in the Configure command, but the result shows **THIRTY_TWO_BIT mode**. I wonder what its meaning and will this made wrong library?

Thanks a lot!

The following shows the command and result:
**\# ./Configure no-shared no-asm solaris64-sparcv9-gcc**
Configuring OpenSSL version 1.1.0c (0x1010003fL)
    no-asan         [default]  OPENSSL_NO_ASAN
    no-asm          [option]   OPENSSL_NO_ASM
    no-crypto-mdebug [default]  OPENSSL_NO_CRYPTO_MDEBUG
    no-crypto-mdebug-backtrace [default]  OPENSSL_NO_CRYPTO_MDEBUG_BACKTRACE
    no-dynamic-engine [forced]  
    no-ec_nistp_64_gcc_128 [default]  OPENSSL_NO_EC_NISTP_64_GCC_128
    no-egd          [default]  OPENSSL_NO_EGD
    no-fuzz-afl     [default]  OPENSSL_NO_FUZZ_AFL
    no-fuzz-libfuzzer [default]  OPENSSL_NO_FUZZ_LIBFUZZER
    no-heartbeats   [default]  OPENSSL_NO_HEARTBEATS
    no-md2          [default]  OPENSSL_NO_MD2 (skip dir)
    no-msan         [default]  OPENSSL_NO_MSAN
    no-rc5          [default]  OPENSSL_NO_RC5 (skip dir)
    no-sctp         [default]  OPENSSL_NO_SCTP
    no-shared       [option]  
    no-ssl-trace    [default]  OPENSSL_NO_SSL_TRACE
    no-ssl3         [default]  OPENSSL_NO_SSL3
    no-ssl3-method  [default]  OPENSSL_NO_SSL3_METHOD
    no-ubsan        [default]  OPENSSL_NO_UBSAN
    no-unit-test    [default]  OPENSSL_NO_UNIT_TEST
    no-weak-ssl-ciphers [default]  OPENSSL_NO_WEAK_SSL_CIPHERS
    no-zlib         [default] 
    no-zlib-dynamic [default] 
Configuring for solaris64-sparcv9-gcc
CC            =gcc
CFLAG         =-m64 -mcpu=ultrasparc -Wall -DB_ENDIAN -DBN_DIV2W -O3 -pthread -DFILIO_H 
SHARED_CFLAG  =-fPIC
DEFINES       =DSO_DLFCN HAVE_DLFCN_H NDEBUG OPENSSL_THREADS OPENSSL_NO_DYNAMIC_ENGINE OPENSSL_PIC
LFLAG         =
PLIB_LFLAG    =
EX_LIBS       =-lresolv -lsocket -lnsl -ldl 
APPS_OBJ      =
CPUID_OBJ     =mem_clr.o
UPLINK_OBJ    =
BN_ASM        =bn_asm.o
EC_ASM        =
DES_ENC       =des_enc.o fcrypt_b.o
AES_ENC       =aes_core.o aes_cbc.o
BF_ENC        =bf_enc.o
CAST_ENC      =c_enc.o
RC4_ENC       =rc4_enc.o rc4_skey.o
RC5_ENC       =rc5_enc.o
MD5_OBJ_ASM   =
SHA1_OBJ_ASM  =
RMD160_OBJ_ASM=
CMLL_ENC      =camellia.o cmll_misc.o cmll_cbc.o
MODES_OBJ     =
PADLOCK_OBJ   =
CHACHA_ENC    =chacha_enc.o
POLY1305_OBJ  =
BLAKE2_OBJ    =
PROCESSOR     =
RANLIB        =ranlib
ARFLAGS       =
PERL          =perl

THIRTY_TWO_BIT mode
BN_LLONG mode
RC4 uses unsigned char

Configured for solaris64-sparcv9-gcc

BTW, I use `elfdump -e libssl.a | grep ELFCLASS32`, and find nothing, so can I say this library is the right 64 bit library?

**Comment by levitte** [comment]
I'm going to venture a guess, without knowing everything about the Sparcv9 architecture, that it's because the architecture features 32-bit words, and I can only assume that the compilers use that as the default `int` size.

I get my information from [The SPARC Architecture Manual v9](https://cr.yp.to/2005-590/sparcv9.pdf), section 4 "Data Formats".

Either way, the parameters `THIRTY_TWO_BIT` and `BN_LLONG` affect the internals of the BIGNUM library, mostly how the bignums are stored and processed.

Is this sufficient information for you?

**Comment by dot-asm** [comment]
THIRTY_TWO_BIT in solaris64 target is not a typo or omission. The underlying reason for sticking with it in 64-bit build is that 32x32 multiplication is the best SPARCv9-compliant processor can do in *either* case, i.e. amount of computational instructions would be same. While on pros side we are liberated from need to maintain separate bignum assembly modules for 32- and 64-bit builds.

As for no-asm. Personally I'd argue that no-asm is becoming incompatible with today standards of security. In sense that assembly code not only provide better performance, but even better security in terms of resistance to various attacks. In other words no-asm usage should be discouraged.

**Comment by dyduyu** [comment]
Thanks a lot for your help @levitte and @dot-asm .
Initially I didn't use no-asm, but I encountered some problems:
1. In solaris x86 platform(solaris 11), when I built 32 bit library(openssl 1.1.0c), the errors came:
crypto/chacha/chacha-x86.s: Assembler messages:
crypto/chacha/chacha-x86.s:484: Error: no such instruction: `pshufb (%eax),%xmm6'
crypto/chacha/chacha-x86.s:497: Error: no such instruction: `pshufb 16(%eax),%xmm6'
crypto/chacha/chacha-x86.s:508: Error: no such instruction: `pshufb (%eax),%xmm7'
crypto/chacha/chacha-x86.s:523: Error: no such instruction: `pshufb 16(%eax),%xmm7'
crypto/chacha/chacha-x86.s:534: Error: no such instruction: `pshufb (%eax),%xmm6'
crypto/chacha/chacha-x86.s:549: Error: no such instruction: `pshufb 16(%eax),%xmm6'
crypto/chacha/chacha-x86.s:559: Error: no such instruction: `pshufb (%eax),%xmm7'
crypto/chacha/chacha-x86.s:572: Error: no such instruction: `pshufb 16(%eax),%xmm7'
crypto/chacha/chacha-x86.s:582: Error: no such instruction: `pshufb (%eax),%xmm6'
crypto/chacha/chacha-x86.s:596: Error: no such instruction: `pshufb 16(%eax),%xmm6'
crypto/chacha/chacha-x86.s:607: Error: no such instruction: `pshufb (%eax),%xmm7'
crypto/chacha/chacha-x86.s:622: Error: no such instruction: `pshufb 16(%eax),%xmm7'
crypto/chacha/chacha-x86.s:633: Error: no such instruction: `pshufb (%eax),%xmm6'
crypto/chacha/chacha-x86.s:648: Error: no such instruction: `pshufb 16(%eax),%xmm6'
crypto/chacha/chacha-x86.s:658: Error: no such instruction: `pshufb (%eax),%xmm7'
crypto/chacha/chacha-x86.s:671: Error: no such instruction: `pshufb 16(%eax),%xmm7'
But when `no-asm` was specified, the errors disappeared.

2. In windows, we used openssl 1.0.1m before, however it crashed on some platforms.
The error is similar with https://mta.openssl.org/pipermail/openssl-dev/2016-May/006795.html.
Our workaround was using no-asm. For this problem, I wonder if 1.1 is solved?

Regards
